File: Daily Coding Problem/problem #31.cpp
/*
The edit distance between two strings refers to the minimum number of character insertions, deletions, and substitutions 
required to change one string to the other. For example, the edit distance between “kitten” and “sitting” is three: 
substitute the “k” for “s”, substitute the “e” for “i”, and append a “g”.

*/

#include<iostream>
#include<string>
#include <algorithm> 
using namespace std;

int findMin(int a , int b , int c){
    int z = (a < b) ? a : b;
    return ((z < c) ? z : c);
}

int editDistance ( string a , string b){
    int arr[a.length()+1][b.length()+1];

    for ( int i=0;i<a.length()+1;i++){
        for (int j=0;j<b.length()+1;j++){
            if(i==0)
                arr[i][j]=j;
            else if (j==0)
                arr[i][j]=i;

            else if (a[i-1]==b[j-1]){
                arr[i][j]=arr[i-1][j-1];
            }
            else if (a[i-1]!=b[j-1]){
                arr[i][j]=findMin(arr[i-1][j-1],arr[i-1][j],arr[i][j-1])+1;
            }
        }
    }


    return arr[a.length()][b.length()];
}
int main(){

    string a,b;
    cout<<"Enter the 1st String: ";
    cin>>a;
    cout<<"Enter the 2nd String: ";
    cin>>b;
    cout<<"The Edit Distance is: "<<editDistance(a,b)<<"\n";

}

File: Daily Coding Problem/problem #31.py



def edit_distance(str1,str2):
    str1_len = len(str1)+1
    str2_len = len(str2)+1

    
    #arr = [[0]*str2_len]*str1_len
    arr=[]
    a_prev=[]
    a=[]
    
    for i in range(str1_len):
        a_prev=a
        a=[]
        for j in range(str2_len):
            
            if(i==0):
                #arr[i][j]=j
                a.append(j)
                
            elif (j==0):
                #arr[i][j]=i
                a.append(i)
            
            elif (str1[i-1]==str2[j-1]):
                #arr[i][j]=arr[i-1][j-1]
                a.append(arr[i-1][j-1])
                

            elif(str1[i-1]!=str2[j-1]):
                #arr[i][j]=min(arr[i-1][j],arr[i-1][j-1],arr[i][j-1])+1
                a.append(min(a[j-1],a_prev[j-1],a_prev[j])+1)
                
        
        arr.append(a)
        

    
    return arr[str1_len-1][str2_len-1]










str1 = raw_input('Enter the 1st String: ')
str2 = raw_input('Enter the 2nd String: ')

print ('The Edit Distance Between 2 string is: '+str(edit_distance(str1,str2)))